File: 05_arrays/05_23_merge_intervals/merge_intervals.cpp
#include <bits/stdc++.h>

using namespace std;

vector<vector<int>> merge(vector<vector<int>> &intervals)
{
    if (intervals.size() <= 1)
        return intervals;
    sort(intervals.begin(), intervals.end());
    vector<vector<int>> result;
    result.push_back(intervals[0]);
    for (int i = 1; i < intervals.size(); i++)
    {
        if (intervals[i][0] > result.back()[1])
        {
            result.push_back(intervals[i]);
        }
        else
        {
            result.back()[1] = max(result.back()[1], intervals[i][1]);
        }
    }
    return result;
}

int main()
{
    vector<vector<int>> arr = {{1, 3}, {8, 10}, {2, 6}, {15, 18}};
    vector<vector<int>> ans = merge(arr);
    cout << "The merged intervals are: "
         << "\n";
    for (auto it : ans)
    {
        cout << "[" << it[0] << ", " << it[1] << "] ";
    }
    cout << endl;
    return 0;
}

File: 05_arrays/05_23_merge_intervals/merge_intervals.py
from typing import List

def merge_intervals(intervals: List[List[int]]):
    n = len(intervals)
    intervals.sort()
    if n == 0:
        return intervals
    result = []
    result.append(intervals[0])
    for i in range(n):
        current = intervals[i]
        last = result[-1]
        if current[0] <= last[1]:
            last[1] = max(last[1], current[1])
        else:
            result.append(current)
    return result

arr = [[1, 3], [8, 10], [2, 6], [15, 18]]
ans = merge_intervals(arr)
print("The merged intervals are:")
for it in ans:
    print(f"[{it[0]}, {it[1]}]", end=" ")
print()